### ISSUE TITLE
Tray icon and buttons glitch after lock

### ISSUE THREAD

**Issue by alvilag** [gnome-pomodoro/gnome-pomodoro#526]
After locking the session, when I unlock it the tray icon and the app buttons look corrupted.

Using Ubuntu 20.04, app installed through Ubuntu Software.

![before](https://user-images.githubusercontent.com/66083813/98259826-de41dd00-1f82-11eb-8a81-4f20dc008564.png)
![after](https://user-images.githubusercontent.com/66083813/98259837-e13ccd80-1f82-11eb-80f2-6bcf42c8d289.png)

**Comment by neumond** [comment]
Extensions page shows error status:
![Screenshot from 2020-11-05 20-40-29](https://user-images.githubusercontent.com/3260024/98276616-486c7900-1fa7-11eb-8b39-2fef58e4cbac.png)

For me, it happens without even session locking, it's enough turn off the screen due to inactivity.

Restarting gnome shell fixes the issue.

Looks like there're many problems with Ubuntu 20.04 described here #489 including tray icon issue. Just realized there must be notifications. Keyboard shortcut also doesn't work.

**Comment by Keyikedalube** [comment]
Same issue on Debian Unstable. Reported bug [here](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=962993)

Sadly, no activity... so can't really pinpoint if it's an upstream bug or a downstream bug :(

**Comment by liamtimms** [comment]
Been like this for some time, maybe 6 months for me? Seems like the bug may be years old: https://github.com/codito/gnome-pomodoro/issues/331

I'm sure PRs are welcome.

**Comment by bbo2adwuff** [comment]
I also see this glitch.

In `journalctl` I get following:

```
Apr 13 10:18:57 arch gnome-shell[888]: Object .Gjs_pomodoro_arun_codito_in_notifications_PomodoroScreenShieldNotification (0x5632c51f5f20), has been already deallocated — impossible to access it. This might be caused by the object having >
Apr 13 10:18:57 arch gnome-shell[888]: == Stack trace for context 0x5632c28b8180 ==
Apr 13 10:18:57 arch gnome-shell[888]: #0   5632c4997580 i   resource:///org/gnome/shell/ui/messageTray.js:775 (14f37ff55420 @ 112)
Apr 13 10:18:57 arch gnome-shell[888]: #1   5632c49974e8 i   /usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/notifications.js:209 (37ae77dc7830 @ 316)
Apr 13 10:18:57 arch gnome-shell[888]: #2   5632c4997428 i   /usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:304 (2c47b249edd0 @ 349)
Apr 13 10:18:57 arch gnome-shell[888]: #3   5632c4997390 i   /usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:138 (37ae77dbf380 @ 372)
Apr 13 10:18:57 arch gnome-shell[888]: #4   5632c49972f8 i   /usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:592 (2c47b249e4c0 @ 113)
Apr 13 10:18:57 arch gnome-shell[888]: #5   7ffe38f4af10 b   resource:///org/gnome/shell/ui/extensionSystem.js:108 (14f37ff4a6f0 @ 395)
Apr 13 10:18:57 arch gnome-shell[888]: #6   5632c4997268 i   resource:///org/gnome/shell/ui/extensionSystem.js:613 (14f37ff42ab0 @ 15)
Apr 13 10:18:57 arch gnome-shell[888]: #7   7ffe38f4b980 b   self-hosted:225 (c4197bf1100 @ 273)
Apr 13 10:18:57 arch gnome-shell[888]: #8   5632c49971d8 i   resource:///org/gnome/shell/ui/extensionSystem.js:612 (14f37ff42b00 @ 98)
Apr 13 10:18:57 arch gnome-shell[888]: #9   5632c4997148 i   resource:///org/gnome/shell/ui/extensionSystem.js:630 (14f37ff42a60 @ 82)
Apr 13 10:18:57 arch gnome-shell[888]: #10   7ffe38f4c3c0 b   self-hosted:844 (c4197b24380 @ 423)
Apr 13 10:18:57 arch gnome-shell[888]: #11   7ffe38f4c4b0 b   resource:///org/gnome/gjs/modules/core/_signals.js:114 (c4197b0f5b0 @ 439)
Apr 13 10:18:57 arch gnome-shell[888]: #12   7ffe38f4cc40 b   resource:///org/gnome/shell/ui/sessionMode.js:195 (1b7b9b5e2d80 @ 284)
Apr 13 10:18:57 arch gnome-shell[888]: #13   5632c4996fe8 i   resource:///org/gnome/shell/ui/sessionMode.js:156 (1b7b9b5e2ec0 @ 40)
Apr 13 10:18:57 arch gnome-shell[888]: #14   5632c4996f30 i   resource:///org/gnome/shell/ui/screenShield.js:590 (1b7b9b5cdb00 @ 208)
Apr 13 10:18:57 arch gnome-shell[888]: #15   5632c4996e98 i   resource:///org/gnome/shell/ui/screenShield.js:316 (1b7b9b5d6330 @ 36)
Apr 13 10:18:57 arch gnome-shell[888]: #16   5632c4996dc8 i   self-hosted:844 (c4197b24380 @ 454)
Apr 13 10:18:57 arch gnome-shell[888]: #17   5632c4996d38 i   resource:///org/gnome/shell/ui/lightbox.js:193 (14f37ff19c90 @ 29)
Apr 13 10:18:57 arch gnome-shell[888]: #18   5632c4996ca8 i   resource:///org/gnome/shell/ui/environment.js:85 (2c47b249c920 @ 98)
Apr 13 10:18:57 arch gnome-shell[888]: #19   5632c4996c18 i   resource:///org/gnome/shell/ui/environment.js:168 (2c47b249c600 @ 14)
Apr 13 10:18:57 arch gnome-shell[888]: JS ERROR: TypeError: notification is null
                                       _onNotificationShow@resource:///org/gnome/shell/ui/messageTray.js:1053:17
                                       showNotification@resource:///org/gnome/shell/ui/messageTray.js:775:18
                                       show@/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/notifications.js:209:25
                                       _updateNotification@/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:304:39
                                       setMode@/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:138:18
                                       disable@/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js:592:23
                                       _callExtensionDisable@resource:///org/gnome/shell/ui/extensionSystem.js:108:32
                                       _disableAllExtensions/<@resource:///org/gnome/shell/ui/extensionSystem.js:613:22
                                       _disableAllExtensions@resource:///org/gnome/shell/ui/extensionSystem.js:612:52
                                       _sessionUpdated@resource:///org/gnome/shell/ui/extensionSystem.js:630:18
                                       _emit@resource:///org/gnome/gjs/modules/core/_signals.js:114:47
                                       _sync@resource:///org/gnome/shell/ui/sessionMode.js:195:14
                                       pushMode@resource:///org/gnome/shell/ui/sessionMode.js:156:14
                                       activate@resource:///org/gnome/shell/ui/screenShield.js:590:34
                                       _onLongLightbox@resource:///org/gnome/shell/ui/screenShield.js:316:18
                                       onComplete@resource:///org/gnome/shell/ui/lightbox.js:193:18
                                       _makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:85:13
                                       _easeActor/<@resource:///org/gnome/shell/ui/environment.js:168:64
```

**Comment by bbo2adwuff** [comment]
In the following line `notification` seems to be empty, therefore `show()` doesn't work, right?

https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js#L304

**Comment by bbo2adwuff** [comment]
To be honest, I didn't have the time to dig into the `ExtensionMode.RESTRICTED` line further...

But as a quick&dirty work around can anyone try replacing the following with the code below?

https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js#L590-L599

```js
    if (extension) {
          extension.destroy();
    }
```

To give it a try just change it in `/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js`, restart and see if the problem still occurs.

**Comment by bbo2adwuff** [comment]
@kamilprusko could it be that in [extension.js](https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js) the [line 303](https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js#L303) leads to `notification = null` and therefore [line 304](https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js#L304) fails?

**Comment by kamilprusko** [comment]
@bbo2adwuff Thanks for debugging the issue.

Your workaround will work. Pushed a151125b98e5bd035683a43d40b17320632023a5 to gnome-4.0 branch, which disables lock screen notification for the time being.

> could it be that in extension.js the line 303 leads to notification = null and therefore line 304 fails?

No, `.connect()` is passive. I think the notification gets externally destroyed during execution of `PomodoroNotification.show`... I will sit on it tomorrow.

**Comment by alvilag** [comment]
> To be honest, I didn't have the time to dig into the `ExtensionMode.RESTRICTED` line further...
> 
> But as a quick&dirty work around can anyone try replacing the following with the code below?
> 
> https://github.com/gnome-pomodoro/gnome-pomodoro/blob/0952054d35e9c9c7073ff8918b82b849b6eccc28/plugins/gnome/extension/extension.js#L590-L599
> 
> ```js
>     if (extension) {
>           extension.destroy();
>     }
> ```
> 
> To give it a try just change it in `/usr/share/gnome-shell/extensions/pomodoro@arun.codito.in/extension.js`, restart and see if the problem still occurs.

@kamilprusko @bbo2adwuff I tried this workaround: it resolved the tray icon glitch; however, the app buttons disappeared completely (see image attached). Also, I noticed that the app resulted deactivated in the gnome extensions built-in list, and it was not possible to switch the button to reactivate it from there. I had to add the app in the startup applications in order for it to automatically launch at startup, but with the inconvenient that it now opens the window--not just the app minimized in the tray-- every time I boot up.

So, for me this workaround resolves some problems but it creates others--perhaps more annoying ones.

Thanks in any case for your attention!

![Screenshot from 2021-04-14 10-22-09](https://user-images.githubusercontent.com/66083813/114681025-0857b200-9d0e-11eb-9626-87a822875c8d.png)

**Comment by bbo2adwuff** [comment]
@kamilprusko I'm using your commit since a few hours and it seems to work perfectly well! Thanks for all you effort you put into this nice piece of software!

@alvilag Did you try what I wrote? Or did you use 
https://github.com/gnome-pomodoro/gnome-pomodoro/commit/a151125b98e5bd035683a43d40b17320632023a5? In any case it is better to try the changes from the commit and not the quick&dirty fix that I proposed!

**Comment by alvilag** [comment]
@bbo2adwuff I tried your quick workaround--I thought it was the same as [a151125](https://github.com/gnome-pomodoro/gnome-pomodoro/commit/a151125b98e5bd035683a43d40b17320632023a5), my bad! I just tried the latter and it works flawlessly. Thanks to you for the reply and to @kamilprusko!

**Comment by kamilprusko** [comment]
Made more fixes in branch [gnome-4.0](https://github.com/gnome-pomodoro/gnome-pomodoro/tree/gnome-4.0). The notification on the lock screen is working again.

**Comment by bbo2adwuff** [comment]
For some days I'm using gnome-pomodoro with the last fixes in the gnome-4.0 branch. On my side the issue is fixed! I'd say this issue can be closed.
